stan/math/rev/meta/rev_matrix_type.hpp

```diff
@@ -0,0 +1,38 @@
+#ifndef STAN_MATH_REV_META_REV_MATRIX_TYPE_HPP
+#define STAN_MATH_REV_META_REV_MATRIX_TYPE_HPP
+
+#include <stan/math/prim/meta/disjunction.hpp>
+#include <stan/math/prim/meta/conjunction.hpp>
+#include <stan/math/prim/meta/require_generics.hpp>
+#include <stan/math/prim/meta/return_type.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/core/var.hpp>
+
+namespace stan {
+
+/**
+ * Determines a return type for a function that accepts given inputs and wants
+ * to return a matrix (or vector or row vector) with given compile time number
+ * of rows and columns.
+ *
+ * Returns `var_value` if any of the inputs are `var_value`, otherwise an Eigen
+ * matrix of `var` or `double`.
+ * @tparam Rows number of compile time rows
+ * @tparam Cols number of compile time columns
+ * @tparam Inputs types of arguments the function accepts
+ */
+template <int Rows, int Cols, typename... Inputs>
+struct rev_matrix_type {
+  using type = std::conditional_t<
+      math::disjunction<math::conjunction<
+          is_var<Inputs>, is_eigen<value_type_t<Inputs>>>...>::value,
+      math::var_value<Eigen::Matrix<double, Rows, Cols>>,
+      Eigen::Matrix<return_type_t<Inputs...>, Rows, Cols>>;
+};
+
+template <int Rows, int Cols, typename... Inputs>
+using rev_matrix_t = typename rev_matrix_type<Rows, Cols, Inputs...>::type;
```

Comment:
Why not `var_matrix_t<>`?

Reply:
In #2050 I have a `is_var_matrix<>` that checks whether a type is either `var<Mat>` or `Mat<var>` so it feels like this is in the same family

Reply:
I could do that, but `var_matrix` sounds to me like it should be `var_value<matrix>`. Maybe you can change your PR instead to be consistent?

Reply:
Yeah I had trouble naming it. `is_rev_matrix<>`, I think that can make sense I'll change mine